### PR TITLE
Fixes 14370: Add Azure Client, support Default Creds

### LINKED ISSUE #14370: Connect to Azure SQL using an Azure AD service principal
**Is your feature request related to a problem? Please describe.**
At the moment it's only possible to connect to an Azure SQL database using a username/password combination. This doesn't seem very secure.

**Describe the solution you'd like**
It would be nice to connect to the database using an Azure AD service principal

**Describe alternatives you've considered**
It might also be possible to use connection arguments see: https://stackoverflow.com/questions/56582031/sqlalchemy-connecting-to-microsoft-azure-active-directory-password

**Additional context**
This might be a further source of inspiration: https://docs.getdbt.com/docs/core/connect-data-platform/mssql-setup#azure-active-directory-authentication-aad

### PR DESCRIPTION
**Pull request by ayush-shah**
### Describe your changes:

Fixes #14370 

Add Mysql, Postgres, AzureSql, Datalake and DBT related connection / config / client changes

#
### Type of change:

- [ ] Bug fix
- [ ] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [ ] I have commented on my code, particularly in hard-to-understand areas. 
- [ ] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

### CODE REVIEW

File: ingestion/tests/unit/metadata/cli/resources/profiler_workflow.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 """
 This file has been generated from dag_runner.j2
 """
-from openmetadata.workflows import workflow_factory
+from openmetadata_managed_apis.workflows import workflow_factory
```
Comment (pmbrull) [inline ingestion/tests/unit/metadata/cli/resources/profiler_workflow.py:4]: oh were these not executed? I guess they'd fail. Maybe we need to review the Makefile

File: openmetadata-spec/src/main/resources/json/schema/entity/services/connections/database/azureSQLConnection.json [hunk 2]
```diff
@@ -59,6 +63,43 @@
...
+        "connectionTimeout": {
+          "title": "Connection Timeout",
+          "description": "Connection Timeout from Connection String for AzureSQL.",
+          "type": "integer",
+          "default": 30
+        }
+      },
+      "required": [
+        "Encrypt",
+        "Trust Server Certificate",
```
Comment (pmbrull) [inline openmetadata-spec/src/main/resources/json/schema/entity/services/connections/database/azureSQLConnection.json:98]: should the required be the title or the name?

File: ingestion/src/metadata/ingestion/source/database/azuresql/connection.py [hunk 3]
```diff
@@ -40,26 +41,43 @@ def get_connection_url(connection: Union[AzureSQLConnection, MssqlConnection]) -
...
+        connection_string = f"Driver={connection.driver};Server={connection.hostPort};Database={connection.database};"
+        connection_string += f"Uid={connection.username};"
+        if (
+            connection.authenticationMode.authentication
+            == Authentication.ActiveDirectoryPassword
+        ):
+            connection_string += f"Pwd={connection.password.get_secret_value()};"
+
+        connection_string += f"Encrypt={'yes' if connection.authenticationMode.encrypt else 'no'};TrustServerCertificate={'yes' if connection.authenticationMode.trustServerCertificate else 'no'};"
+        connection_string += f"Connection Timeout={connection.authenticationMode.connectionTimeout or 30};Authentication={connection.authenticationMode.authentication.value};"
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/database/azuresql/connection.py:54]: would be good to have a test validating how we're building these strings

File: ingestion/src/metadata/ingestion/source/database/mysql/connection.py [hunk 4]
```diff
@@ -38,6 +42,14 @@ def get_connection(connection: MysqlConnection) -> Engine:
     """
     Create connection
     """
+    if hasattr(connection.authType, "azureConfig"):
+        azure_client = AzureClient(connection.authType.azureConfig).create_client()
+        if not connection.authType.azureConfig.scopes:
+            raise ValueError("Azure Scopes are not provided")
```
Comment (pmbrull) [inline ingestion/src/metadata/ingestion/source/database/mysql/connection.py:48]: if they are required, should we mark them as required in the JSON Schema? 🤔 I might be missing something

Comment (pmbrull) [reply ingestion/src/metadata/ingestion/source/database/mysql/connection.py:48]: and I guess we'll followup this PR with the docs right

**Review by pmbrull — DISMISSED** [review]

File: ingestion/tests/unit/metadata/cli/resources/profiler_workflow.py (hunk 1 above)
Comment (ayush-shah) [reply ingestion/tests/unit/metadata/cli/resources/profiler_workflow.py:4]: Yes, it should have failed, not sure why it didn't.. didn't take a closer look, fixed the imports though

File: openmetadata-spec/src/main/resources/json/schema/entity/services/connections/database/azureSQLConnection.json (hunk 2 above)
Comment (ayush-shah) [reply openmetadata-spec/src/main/resources/json/schema/entity/services/connections/database/azureSQLConnection.json:98]: ahh, right, my bad.. but i tried with the required using name, i think it's not working as expected, the Authentication if not provided with anything, should be optional, if we add ActiveDirectory related type, then it should have other inputs added by the user, removing the same..
Screenshot 2024-03-21 at 9 32 50 PM
As all the params will be available from azure itself, the users will have to copy paste it here.. and some of the params are not needed somewhere

File: ingestion/src/metadata/ingestion/source/database/azuresql/connection.py (hunk 3 above)
Comment (ayush-shah) [reply ingestion/src/metadata/ingestion/source/database/azuresql/connection.py:54]: Will add the same, was waiting for sonar to pick it up which of these we can add tests for.

File: ingestion/src/metadata/ingestion/source/database/mysql/connection.py (hunk 4 above)
Comment (ayush-shah) [reply ingestion/src/metadata/ingestion/source/database/mysql/connection.py:48]: Scopes are just needed for Postgresql and Mysql Azure Auth, not at places like Datalake / ADLS

**Review by pmbrull — APPROVED** [review]

**Pull request merged**